app/controllers/api/custom_buttons_controller.rb

```diff
@@ -2,32 +2,32 @@ module Api
   class CustomButtonsController < BaseController
     def create_resource(_type, _id, data)
       CustomButton.transaction do
-        custom_button = CustomButton.new(data.except("resource_action", "options"))
-        custom_button.userid = User.current_user.userid
-        custom_button.options = data["options"].deep_symbolize_keys if data["options"]
-        custom_button.resource_action.update!(data["resource_action"].deep_symbolize_keys) if data.key?("resource_action")
-        custom_button.save!
-        custom_button
+        CustomButton.new(data.except("resource_action", "options")).tap do |cb|
+          cb.userid = User.current_user.userid
+          cb.options = data["options"].deep_symbolize_keys if data["options"]
+          cb.create_resource_action!(data["resource_action"].deep_symbolize_keys) if data.key?("resource_action")
+          cb.save!
+        end
       end
-    rescue
-      raise BadRequestError, "Failed to create new custom button - #{custom_button.errors.full_messages.join(", ")}"
+    rescue => err
+      raise BadRequestError, "Failed to create new custom button - #{err.message}"
     end
 
     def edit_resource(type, id, data)
       return if data.empty?
 
       CustomButton.transaction do
-        custom_button = fetch_custom_button(type, id)
-        updated_data = data['resource'] || data
-        if updated_data['resource_action'].present?
-          custom_button.resource_action = create_or_update_resource_action(updated_data['resource_action'])
+        fetch_custom_button(type, id).tap do |cb|
+          updated_data = data['resource'] || data
+          if (resource_action = updated_data.delete("resource_action")).present?
```

Comment:
should this be `updated_data = (data['resource'] || data).dup` or `updated_data = data['resource'] || data.dup` since `updated_data.delete` is being called on the next line?

Reply:
you could dup or you could use an except() in deep_symbolize_keys.

Before, we had updated_data.present? so we needed to modify the hash to get present correct.
Now, I think we may be able to avoid this

Reply:
looks like we need updated_data to not have 'resource_action' in it (aka drew was right)
So lets follow oleg's suggestion to dup the hash